DOCS-886: Improve Go modular resource code template by adding module with contextual main (#1449)
viamrobotics/docs: docs/extend/modular-resources/_index.md

{{< tabs name="Base Model Modules" >}}
{{% tab name="Go" %}}

> The example custom model code (<file>mybase.go</file>) and module (<file>mymodule.go</file>) are adapted from the full demo module available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/models/mybase/mybase.go), and creates a modular resource implementing Viam's built-in Base API (rdk:service:base).

Author: ```suggestion
The code for this custom model (<file>mybase.go</file>) and module (<file>mymodule.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/models/mybase/mybase.go).
The module registers a modular resource implementing Viam's built-in Base API (rdk:service:base).
```

> This example custom model code (<file>mybase.go</file>) and module (<file>mymodule.go</file>) is adapted from the full demo module available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/models/mybase/mybase.go), and creates a singular modular resource implementing Viam's built-in Base API (rdk:service:base).

Reviewer: ```suggestion
The example custom model code (<file>mybase.go</file>) and module (<file>mymodule.go</file>) are adapted from the full demo module available on the [Viam GitHub repository](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/models/mybase/mybase.go), and creates a singular modular resource implementing Viam's built-in Base API (rdk:service:base).
```

supernit: The above aligns plural (`the ... code ... and module are adapted`) and article (`the ... repo`). The article change also applicable to the Python tab, if you like!

> This example custom model code (<file>mybase.go</file>) and module (<file>mymodule.go</file>) is adapted from the full demo module available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/models/mybase/mybase.go), and creates a singular modular resource implementing Viam's built-in Base API (rdk:service:base).

Reviewer 2: I recognize that my comment is about a part of the line you didn't edit, but what does "singular" mean in this context? It's not the only one, it's not the best one, it's not a unique one. Could we remove that word and just say it "creates a modular resource"?

}
```

<file>mymodule.go</file>

``` go {class="line-numbers linkable-line-numbers"}
// Package main is a module which serves the mybase custom model type.
package main

import (
    "context"

    "github.com/edaniels/golog"

    "go.viam.com/rdk/components/base"
    "go.viam.com/rdk/module"
    "go.viam.com/utils"

>     // NOTE: You must update the following line to import your local package "mybase"

Reviewer 2: :+1: Excellent comment here!

>     // NOTE: You must update the following line to import your local package "mybase"

Author: @penguinland LMK if this does not provide enough info here-- would this be the path to mybase after compiling?

Reviewer 2: "After compiling" doesn't really make sense. It's the path to mybase, full stop. I like the comment: it's saying "you probably changed the package name when you made your own version of the code, so remember to change this line to match."

    "go.viam.com/rdk/examples/customresources/models/mybase"
)

func main() {
    // NewLoggerFromArgs will create a golog.Logger at "DebugLevel" if

>     // "--log-level=debug" is the third argument in os.Args and at "InfoLevel"

Reviewer 2: Not sure about this "third argument" part. Shouldn't the argument flags work in any order? Hopefully it'll create the logger at the right debug level if that arg is present anywhere in `os.Args`.

Author: @penguinland  I adapted this sample from the RDK demos and this note came w/ --> I agree it seems misleading if the flag is what's triggering 
Making this change here

    // otherwise.

>     utils.ContextualMain(mainWithArgs, module.NewLoggerFromArgs("complexmodule"))

Author: ```suggestion
    utils.ContextualMain(mainWithArgs, module.NewLoggerFromArgs("yourmodule"))
```

Your options for completing this step are flexible, as this file does not need to be in a raw binary format.

If using the Go SDK, Go will build a binary when you compile your module.

> Expand the [Go module examples](#code-your-module) to view <file>mymodule.go</file> as an example of a module you can run that initializes a custom model <file>myBase.go</file> from the registry.

Author: ```suggestion
Expand the [Go module examples](#code-your-module) to view <file>mymodule.go</file> as an example of a module you can run that initializes a custom model <file>mybase.go</file> from the registry.
```
